cli/ops/net.rs

```diff
@@ -140,88 +214,206 @@ fn op_receive(
   Ok(JsonOp::Async(op.boxed_local()))
 }
 
-#[derive(Deserialize)]
-struct SendArgs {
-  rid: i32,
-  hostname: String,
-  port: u16,
-  transport: String,
-}
-
-fn op_send(
+#[cfg(unix)]
+fn receive_unix_packet(
   state: &State,
-  args: Value,
+  args: ReceiveArgs,
   zero_copy: Option<ZeroCopyBuf>,
 ) -> Result<JsonOp, OpError> {
-  assert!(zero_copy.is_some());
-  let buf = zero_copy.unwrap();
+  let mut buf = zero_copy.unwrap();
 
-  let args: SendArgs = serde_json::from_value(args)?;
-  assert_eq!(args.transport, "udp");
   let rid = args.rid as u32;
 
   let state_ = state.clone();
-  state.check_net(&args.hostname, args.port)?;
 
   let op = async move {
     let mut state = state_.borrow_mut();
     let resource = state
       .resource_table
-      .get_mut::<UdpSocketResource>(rid)
+      .get_mut::<UnixDatagramResource>(rid)
       .ok_or_else(|| OpError::other("Socket has been closed".to_string()))?;
-
-    let socket = &mut resource.socket;
-    let addr = resolve_addr(&args.hostname, args.port).await?;
-    socket.send_to(&buf, addr).await?;
-
-    Ok(json!({}))
+    let (size, remote_addr) = resource.socket.recv_from(&mut buf).await?;
+    Ok(json!({
+      "size": size,
+      "remoteAddr": {
+        "address": remote_addr.as_pathname(),
+        "transport": args.transport,
+      }
+    }))
   };
 
   Ok(JsonOp::Async(op.boxed_local()))
 }
 
+fn op_receive(
+  state: &State,
+  args: Value,
+  zero_copy: Option<ZeroCopyBuf>,
+) -> Result<JsonOp, OpError> {
+  assert!(zero_copy.is_some());
+  let args: ReceiveArgs = serde_json::from_value(args)?;
+  match args.transport.as_str() {
+    "udp" => receive_udp(state, args, zero_copy),
+    #[cfg(unix)]
+    "unixpacket" => receive_unix_packet(state, args, zero_copy),
+    _ => Err(OpError::other(format!(
+      "Unsupported transport protocol {}",
+      args.transport
+    ))),
+  }
+}
+
 #[derive(Deserialize)]
-struct ConnectArgs {
+struct SendArgs {
+  rid: i32,
   transport: String,
-  hostname: String,
-  port: u16,
+  #[serde(flatten)]
+  transport_args: ArgsEnum,
 }
 
-fn op_connect(
+fn op_send(
   state: &State,
   args: Value,
-  _zero_copy: Option<ZeroCopyBuf>,
+  zero_copy: Option<ZeroCopyBuf>,
 ) -> Result<JsonOp, OpError> {
-  let args: ConnectArgs = serde_json::from_value(args)?;
-  assert_eq!(args.transport, "tcp"); // TODO Support others.
+  assert!(zero_copy.is_some());
+  let buf = zero_copy.unwrap();
   let state_ = state.clone();
-  state.check_net(&args.hostname, args.port)?;
+  match serde_json::from_value(args)? {
+    SendArgs {
+      rid,
+      transport,
+      transport_args: ArgsEnum::Ip(args),
+    } if transport == "udp" => {
+      state.check_net(&args.hostname, args.port)?;
```

Comment:
I'm having trouble mentally parsing this rust code... Is there a way to rewrite this this match?